tbot_contrib/locking.py

```diff
@@ -0,0 +1,302 @@
+                repeat = False  # We got a lock, we're done
+            else:  # No lock this time
+                # Read the lock, perhaps it is stale... This may race!
+                retval, result = labhost.exec("cat", lockfile)
+                if retval == 0:
+                    results = result.split()
+                    # Check if the lock can expire, and whether it has expired.
+                    if (int(results[0]) != 0) and (int(results[0]) < int(time.time())):
+                        # Check whether the lock PID is still in use by a bash instance
+                        retval, result = labhost.exec("ps", "-ocomm=", results[1])
+                        if (retval == 0) and (result.strip() == "bash"):
+                            print(f"Expired lock {lockfile} PID still appears valid!")
+                        else:
+                            labhost.exec("rm", lockfile)  # Delete stale lock
+                            print(f"Lock {lockfile} expired, deleted.")
+                            repeat = (
+                                True  # The stale lock may be available next time round.
+                            )
+            if (
+                success
+            ):  # Because things could race, test whether we still have the lock...
+                time.sleep(self.lock_sleep)  # Allow things to settle
+                # Check whether the lockfile still exists and get its inode.
+                retval, result = labhost.exec("stat", "-c%i", lockfile)
+                if retval == 0:  # lock still exists
+                    # Get inode of the tempfile, to see if it matches lockfile's
+                    if result != labhost.exec("stat", "-c%i", tempfile)[1]:
+                        repeat = True  # Inodes don't match, try again
+                else:
+                    repeat = True  # our lock got deleted, try again.
+            elif repeat:  # No lock, but we may repeat
+                print("Sleep..")
+                time.sleep(self.lock_sleep)  # Back-off to let things settle
+
+        labhost.exec("rm", tempfile)  # Delete the temp file
+        return success
+
+    @contextlib.contextmanager
+    def _init_pre_connect(self) -> typing.Iterator:
+        with tbot.ctx.request(tbot.role.LabHost) as labhost:
+            self.lock_no = None
+            for count in range(self.lock_pool_max):  # iterate through locks
+                self.lock_file = linux.Path(labhost, f"{self.lock_name}{count}")
+                print(f"Try locking on {count} using {self.lock_file}...")
+                if self._locker(labhost, self.lock_file):
+                    self.lock_no = count
+                    break
+            if self.lock_no is None:
+                raise Exception("Could not get free lock")
+            try:
+                yield None
+            finally:
+                labhost.exec("rm", "-f", self.lock_file)
+
+
+class MachineLock2(machine.PreConnectInitializer):
```

Comment:
Look, the only reason the second implementation exists, is because it does what I understood the spec to request. The problem is, if a lock "expires" but the previous "owner" is still using it, the lock would get re-used without the "owner" knowing about it, which creates issues which are beautifully illustrated by multiple Sandboxes; the 3rd implementation using the "lockfile" utility suffers from exactly the same issue. The idea for storing the PID comes from the 4th implementation, which in turn does not expire.

So the goal is to only keep "first" implementation. The other implementations remain only as references for now, partly to show that multiple implementations can co-exist. They are also duplicated so that modifying one doesn't break the others.

Reply:
Ah, okay, I was under the impression you want to keep both.  No problem then!